Tests/UnitTests/RequestCoordinatorTests.swift

```diff
@@ -0,0 +1,249 @@
+    var coordinator: RequestCoordinator!
+    
+    override func setUp() {
+        session = TestURLSession()
+        coordinator = RequestCoordinator(session: session)
+    }
+    
+    func testInitialization() {
+        XCTAssertEqualReferences(coordinator.session as AnyObject, session)
+    }
+    
+    func testEnqueueRetriableRequest_createsDataTask() {
+        // Arrange
+        let request = URLRequest(url: URL(string: "https://usebutton.com")!)
+        
+        // Act
+        coordinator.enqueueRetriableRequest(request: request,
+                                            attempt: 0,
+                                            maxRetries: 0,
+                                            retryIntervalInMS: 0) { _, _ in }
+        
+        // Assert
+        XCTAssertTrue(session.didCallDataTaskWithRequest)
+        XCTAssertNotNil(session.lastDataTask)
+        XCTAssertEqual(session.lastDataTask?.originalRequest, request)
+        XCTAssertTrue(session.lastDataTask!.didCallResume)
+        XCTAssertEqual(session.allDataTasks.count, 1)
+    }
+    
+    func testEnqueueRetriableRequest_success() {
+        // Arrange
+        let expectation = self.expectation(description: "success")
+        let url = URL(string: "https://usebutton.com")!
+        let expectedData = Data()
+        
+        // Act
+        coordinator.enqueueRetriableRequest(request: URLRequest(url: url), attempt: 0, maxRetries: 0, retryIntervalInMS: 0) { data, error in
+            
+            // Assert
+            XCTAssertNil(error)
+            XCTAssertEqual(data, expectedData)
+            XCTAssertEqual(self.session.allDataTasks.count, 1)
+            
+            expectation.fulfill()
+        }
+        
+        let response = HTTPURLResponse(url: url, statusCode: 200, httpVersion: nil, headerFields: nil)
+        session.lastDataTask?.completion(expectedData, response, nil)
+        
+        self.wait(for: [expectation], timeout: 2.0)
+    }
+    
+    func testEnqueueRetriableRequest_failure() {
+        // Arrange
+        let expectation = self.expectation(description: "failure")
+        let url = URL(string: "https://usebutton.com")!
+                let expectedError = TestError.known
+        
+        // Act
+        coordinator.enqueueRetriableRequest(request: URLRequest(url: url), attempt: 0, maxRetries: 0, retryIntervalInMS: 0) { data, error in
+            
+            // Assert
+            XCTAssertNil(data)
+            XCTAssertEqual(error as? TestError, expectedError)
+            XCTAssertEqual(self.session.allDataTasks.count, 1)
+            
+            expectation.fulfill()
+        }
+        
+        let response = HTTPURLResponse(url: url, statusCode: 500, httpVersion: nil, headerFields: nil)
+        session.lastDataTask?.completion(nil, response, expectedError)
+        
+        self.wait(for: [expectation], timeout: 2.0)
+    }
+    
+    func testEnqueueRetriableRequest_noData_retries() {
```

Comment:
Why would we retry on no data?

Reply:
In the case of no network connection, there is no data or response